### PR TITLE
Add api usage snippets

### PR DESCRIPTION
**Pull request by jack-berg**
Code snippets support opentelemetry.io PR: https://github.com/open-telemetry/opentelemetry.io/pull/5276

### CODE REVIEW

File: doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    AtomicLong widgetCount = new AtomicLong();
+
+    // Construct an async counter to observe an existing counter in a callback
+    ObservableLongCounter asyncCounter =
+        meter
+            .counterBuilder("fully.qualified.counter")
+            .setDescription("A count of produced widgets")
+            .setUnit("{widget}")
+            // optionally change the type to double
+            // .ofDoubles()
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java:27]:
```suggestion
            // type can be changed to double by calling .ofDoubles()
```

File: doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            // the callback is invoked a MetricReader reads metrics
+            .buildWithCallback(
+                observableMeasurement -> {
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java:30]:
```suggestion
            .buildWithCallback(
                // the callback is invoked when a MetricReader reads metrics
                observableMeasurement -> {
```

File: doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    AtomicReference<Double> processingLineTemp = new AtomicReference<>(273.0);
+
+    // Construct an async counter to observe an existing counter in a callback
+    ObservableDoubleGauge asyncGauge =
+        meter
+            .gaugeBuilder("fully.qualified.gauge")
+            .setDescription("The current temperature of the widget processing line")
+            .setUnit("K")
+            // optionally change the type to long
+            // .ofLongs()
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java:26]:
```suggestion
            // type can be changed to long by calling .ofLongs()
```

File: doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            // the callback is invoked a MetricReader reads metrics
+            .buildWithCallback(
+                observableMeasurement -> {
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java:29]:
```suggestion
            .buildWithCallback(
                // the callback is invoked when MetricReader reads metrics
                observableMeasurement -> {
```

**Review by trask — COMMENTED** [review]
a couple of comments about comments (if you accept them, there are several duplicates throughout)

File: doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java (hunk 3 above)
Comment (jack-berg) [reply doc-snippets/api/src/main/java/otel/AsyncGaugeUsage.java:26]: Same comment here: https://github.com/open-telemetry/opentelemetry-java-examples/pull/498#discussion_r1783475243

File: doc-snippets/api/src/main/java/otel/SpanUsage.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            // be set using Context.current()
+            // .setParent(parentContext)
+            // Optionally add links
+            // .addLink(linkContext, linkAttributes)
+            // Start the span
+            .startSpan();
+
+    // Check if span is recording before computing additional data
+    if (span.isRecording()) {
+      // Update the span name with information not avilable when starting
```
Comment (jaydeluca) [inline doc-snippets/api/src/main/java/otel/SpanUsage.java:55]:
```suggestion
      // Update the span name with information not available when starting
```

**Review by jaydeluca — APPROVED** [review]
reviewing this was fun, thanks for the tour of the api! I think these are really useful examples and the added commentary is great

File: doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java (hunk 1 above)
Comment (trask) [reply doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java:27]: makes sense, what do you think of this?

```suggestion
            // uncomment to optionally change the type to double
            // .ofDoubles()
```

Comment (jack-berg) [reply doc-snippets/api/src/main/java/otel/AsyncCounterUsage.java:27]: I like it. I'll update it elsewhere too.

File: doc-snippets/api/src/main/java/otel/LogRecordUsage.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            Arrays.asList("value1", "value2"))
+        .setAttribute(
+            AttributeKey.booleanArrayKey("come.acme.bool-array-key"), Arrays.asList(true, false))
+        .setAttribute(AttributeKey.longArrayKey("come.acme.long-array-key"), Arrays.asList(1L, 2L))
+        .setAttribute(
+            AttributeKey.doubleArrayKey("come.acme.double-array-key"), Arrays.asList(1.1, 2.2))
+        .setAllAttributes(WIDGET_RED_CIRCLE)
+        // Uncomment to optionally explicitly set the context used to correlate with spans. If
+        // omitted,
+        // Context.current() is used.
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/LogRecordUsage.java:59]:
```suggestion
        // omitted, Context.current() is used.
```

File: doc-snippets/api/src/main/java/otel/SpanUsage.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                AttributeKey.doubleArrayKey("come.acme.double-array-key"), Arrays.asList(1.1, 2.2))
+            // Optionally omit initializing AttributeKey
+            .setAttribute("com.acme.string-key", "value")
+            .setAttribute("com.acme.bool-key", true)
+            .setAttribute("come.acme.long-key", 1L)
+            .setAttribute("come.acme.double-key", 1.1)
+            .setAllAttributes(WIDGET_RED_CIRCLE)
+            // Uncomment to optionally explicitly set the parent span context. If omitted, the
+            // span's parent will
+            // be set using Context.current()
```
Comment (trask) [inline doc-snippets/api/src/main/java/otel/SpanUsage.java:47]:
```suggestion
            // span's parent will be set using Context.current()
```

**Review by trask — APPROVED** [review]

**Pull request merged**